modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportUpdateByQueryAction.java

```diff
@@ -78,14 +82,33 @@ protected void doExecute(UpdateByQueryRequest request, ActionListener<BulkIndexB
      * Simple implementation of update-by-query using scrolling and bulk.
      */
     static class AsyncIndexBySearchAction extends AbstractAsyncBulkIndexByScrollAction<UpdateByQueryRequest> {
+
+        private final ScriptApplier scriptApplier;
+
         public AsyncIndexBySearchAction(BulkByScrollTask task, ESLogger logger, ScriptService scriptService,
                 ParentTaskAssigningClient client, ThreadPool threadPool, ClusterState clusterState, UpdateByQueryRequest request,
                 ActionListener<BulkIndexByScrollResponse> listener) {
-            super(task, logger, scriptService, clusterState, client, threadPool, request, request.getSearchRequest(), listener);
+            super(task, logger, client, threadPool, request, request.getSearchRequest(), listener);
+
+            Script script = mainRequest.getScript();
+            if (script != null) {
+                this.scriptApplier = new UpdateByQueryScriptApplier(task, scriptService, script, clusterState, script.getParams());
+            } else {
+                this.scriptApplier = null;
+            }
         }
 
         @Override
-        protected IndexRequest buildIndexRequest(SearchHit doc) {
+        public ActionRequest<?> apply(ActionRequest<?> request, SearchHit doc) {
+            ActionRequest<?> result = super.apply(request, doc);
+            if (scriptApplier == null) {
```

Comment:
This feels like something the superclass should handle. Reindex and update-by-query can pass the applier into the async action and delete-by-query can pass null? Then the superclass doesn't have to implement BiFunction?